_gsocblogs/2025/GSoC Final Evaluation_Prasanna Kasar.md

```diff
@@ -0,0 +1,189 @@
+
+## Details
+
+|  |  |
+| --- | --- |
+| Name | [Prasanna Kasar](https://github.com/prasannakasar) |
+| Organisation | [CERN HSF (Root Project)](https://github.com/root-project/root) |
+| Mentor | [Sanjiban Sengupta](https://github.com/sanjibansg), [Dr. Lorenzo Moneta](https://github.com/lmoneta)|
+| Project | [TMVA SOFIE - Enhancing Keras Parser and JAX/FLAX Integration](https://summerofcode.withgoogle.com/programs/2025/projects/uAjGYhgX) |
+
+## Project Description
+
+The SOFIE (System for Optimized Fast Inference Code Emit) project is an initiative within the TMVA (Toolkit for Multivariate Data Analysis) framework in ROOT, which aims to enhance the efficiency and speed of inference for machine learning models. SOFIE converts ML models trained in different frameworks, such as ONNX, PyTorch, and TensorFlow, into an Intermediate Representation (IR). This IR allows SOFIE to generate optimized C++ functions for fast and effective inference of neural networks and subsequently convert them into C++ header files, which can be used in a plug-and-go style for inference.
+
+## SOFIE's workflow
+
+To reduce the overhead of using multiple frameworks for inference, SOFIE generates unified inference code for models trained with different frameworks.
+
+<img width="512" height="235" alt="image" src="https://gist.github.com/user-attachments/assets/bf1f9c4d-28e4-46c0-bdff-ab5653960512" />
+
+SOFIE has mainly two components: a Parser and an inference code generator.
+
+<img width="867" height="274" alt="image" src="https://gist.github.com/user-attachments/assets/096f2af8-72d4-4551-8fd5-4208f3ed1894" />
+
+SOFIE currently supports parsing mechanisms for ML models built with frameworks such as ONNX, PyTorch, and TensorFlow.
+
+## About SOFIE's Keras Parser
+
+Currently, SOFIE's existing Keras parser is written in C++ and is quite old. Although it's written in C++, the actual parsing logic is implemented in Python. Additionally, it lacks support for parsing layers such as Pooling and LayerNormalization.
+
+## Project Objectives
+
+- Rewrite the Keras model parser in Python, replacing earlier C++ logic to improve modular design and flexibility, and simplify future extensions  
+- Extend parser functionality to support Pooling and LayerNormalization layers  
+- Enable support for Keras 3, while preserving support for Keras 2.x models, ensuring full backward compatibility  
+- Add support for both types of models, i.e., models built using Keras' Functional as well as Sequential API.  
+- Design comprehensive unit tests for the parser to guarantee robustness and correctness  
+
+## Work Accomplished
+
+Since SOFIE's operators are written entirely in C++, we had to leverage ROOT's `Pythonization` functionality, which essentially allows us to use SOFIE's C++ objects in a Pythonic interface. The overall structure of the parser is very similar to the previous one. The sequence of operations is as follows:
```

Comment:
```suggestion
Since SOFIE's operators are written entirely in C++, we had to leverage ROOT's `Pythonization` functionality, which essentially allows us to bind SOFIE's C++ objects to a Pythonic interface. The overall structure of the parser is very similar to the previous one. The sequence of operations is as follows:
```